fs/zipfs.py

```diff
@@ -35,16 +35,21 @@ def __init__(self, fs, name):
     def read(self, size=-1):
         if self._pos >= self._end:
             return b''
-        elif size is None or size < 0:
+        if size is None or size < 0:
             size = self._end - self._pos
-            # NB(@althonos): do NOT replace by self._f.read() !
-            buf = self._f.read(size-1) + self._f._readbuffer[-1:]
-            self._f._offset += 1
-        elif self._f._offset + size <= len(self._f._readbuffer):
+        if self._f._offset + size <= len(self._f._readbuffer):
             buf = self._f._readbuffer[self._f._offset:size+self._f._offset]
             self._f._offset += size
-        else:
+        elif size < self._end - self._pos < self._f.MIN_READ_SIZE:
             buf = self._f.read(size)
+        elif size < self._f.MIN_READ_SIZE:
+            buf = self._f.read(size-1) + self._f._readbuffer[-1:]
+        else:
+            buf = bytearray()
```

Comment:
Shouldn't this method return a `bytes` object? Could be wrong, but I don't believe that `bytearray.extend` is more efficient that just adding strings. You're probably better of doing a `b''.join( ... )` here.